src/pip/_internal/download.py

```diff
@@ -954,6 +956,50 @@ def unpack_http_url(
             os.unlink(from_path)
 
 
+def _copy2_ignoring_special_files(src, dest):
+    # type: (str, str) -> None
+    """Copying special files is not supported, but as a convenience to users
+    we skip errors copying them. This supports tools that may create e.g.
+    socket files in the project source directory.
+    """
+    try:
+        copy2_fixed(src, dest)
+    except shutil.SpecialFileError as e:
+        # SpecialFileError may be raised due to either the source or
+        # destination. If the destination was the cause then we would actually
+        # care, but since the destination directory is deleted prior to
+        # copy we ignore all of them assuming it is caused by the source.
+        logger.warning(
+            "Ignoring special file error '%s' encountered copying %s to %s.",
+            str(e),
+            path_to_display(src),
+            path_to_display(dest),
+        )
+
+
+def _copy_source_tree(source, target):
+    # type: (str, str) -> None
+    def ignore(d, names):
+        # Pulling in those directories can potentially be very slow,
+        # exclude the following directories if they appear in the top
+        # level dir (and only it).
+        # See discussion at https://github.com/pypa/pip/pull/6770
+        return ['.tox', '.nox'] if d == source else []
+
+    if PY2:
```

Comment:
If it were me, I would vary only the kwargs, but that's just my preference. I.e. something like:

```python
kwargs = {}
if PY3:
    kwargs.update(copy_function=_copy2_ignoring_special_files)
```

Reply:
Updated [here](https://github.com/pypa/pip/pull/6844/files#diff-34481ab2fefef95e8820a46e86004ae7R1012-R1019). I think I was avoiding this because of the typing, but it ended up not being too bad. mypy doesn't support `kwargs.update(key=value)` (python/mypy#6019), so I went with `kwargs['key'] = value`.